### ISSUE TITLE
Using .. list-table:: creates error on generating Latex pdf

### ISSUE THREAD

**Issue by nmat** [sphinx-doc/sphinx#2593]
To reproduce:

``` rest
Some paragraph text

.. list-table::

  * - .. figure:: /images/testimage01.png

    Test image01

   - .. figure:: /images/testimage02.png

   Test image02

Another paragraph
```

The reason I have set this up in this way is to be able to have 2 images on the same line. Right next to each other. Generating the html works but pdf gives the following error:

`! LaTeX Error: Not in outer par mode.`

`See the LaTeX manual or LaTeX Companion for explanation.`

`l.280 \hline\end{tabulary}`

I am using MikTex also with 1.4.2 of sphinx

**Comment by nmat** [comment]
http://www.tex.ac.uk/FAQ-ouparmd.html

**Comment by tk0miya** [comment]
I'll take a look

**Comment by tk0miya** [comment]
In my short investigation:

Surely it comes from a figure inside a table.
It was rendered as following LaTeX code:

```
\begin{tabulary}{\linewidth}{|L|L|}
\hline
\begin{figure}[htbp]
\centering
\capstart

\includegraphics{{logo}.jpg}
\caption{Test image01}\label{index:id1}\end{figure}
 & \begin{figure}[htbp]
\centering
\capstart

\includegraphics{{logo}.jpg}
\caption{Test image02}\label{index:id2}\end{figure}
\\
\hline\end{tabulary}
```

In my poor LaTeX knowledge, using `minipage` instead figure environment is better way to avoid the error.
But, to use minipage, we have to measure the images as LaTeX units (cf. in, pt, bp, etc.).
I will continue investigation to fix this problem.

Thanks,

@jfbu Could you give me advice?

**Comment by nmat** [comment]
or maybe have a better way to align images in a table like format?

**Comment by jfbu** [comment]
It is not possible to use LaTeX figure environment inside a table. However I found a work-around, somewhat heuristically. It is quite complicated due to other things done by Sphinx like using `\capstart` to get the hyperlinks to point to top of picture, not to caption. But the hardest is that `tabulary` makes two passes. I have a heuristic way to detect if we are in first pass, hence I can inhibit counter stepping for figures from `\caption`. Here is my working patch. Put this in `conf.py`:

``` python
'preamble': 
"""
\\usepackage{etoolbox}
\\makeatletter
\\AtBeginEnvironment{tabulary}
{\\renewenvironment{figure}[1][]% swallow [htbp] 
   {\\let\\captionold\\caption\\def\\@captype{figure}%
    \\def\\caption{\\ifx\\equation$% $ trick to know if first tabulary pass 
                 \\expandafter\\@gobble\\else\\expandafter\\captionold\\fi}%
    \\minipage{\\linewidth}% needed else \\vskip in tabular cell causes error
   }{\\endminipage}}\\makeatother
""",
```

(in `latex_elements`). Then following test reST source works:

``` rst
.. raw:: latex

   \listoffigures

Some paragraph text

.. list-table::

   * - .. figure:: /images/testimage01.png

          Test image01

     - .. figure:: /images/testimage02.png

          Test image02

   * - .. figure:: /images/testimage03.png

          Test image03

     - .. figure:: /images/testimage04.png

          Test image04

Another paragraph

.. figure:: /images/testimage05.png

   Test image05

.. figure:: /images/testimage06.png

   Test image06
```

I added `\listoffigures` to LateX output only to check if hyperlinks are ok, and miraculously they are.

Problem is that Sphinx tables are not always rendered using `tabulary`, one would need to patch other things possibly. LaTeX writer could possibly avoid using figure environment inside tables, but then there is the problem how to do captions, and all table-like environments typesetting more than once their contents will cause issues. I was confronted with that when I tried to improve footnotes, which I have not finished. This is when I detected the trick `\ifx\equation$` to check if in first pass of `tabulary`.

**Comment by jfbu** [comment]
@tk0miya I forgot to say you can't put a `figure` environement inside a `minipage` either. This is why my patch redefines `figure` environment. It uses `minipage` because of `\caption` which does things which are not possible in a table cell, if not hidden in `minipage`.

**Comment by jfbu** [comment]
This might be better:

``` python
'preamble': 
"""
\\usepackage{etoolbox}
\\makeatletter
\\AtBeginEnvironment{tabulary}
{\\renewenvironment{figure}[1][]% swallow [htbp] 
   {\\let\\captionold\\caption\\def\\@captype{figure}%
    \\def\\caption{\\ifx\\equation$% $ trick to know if first tabulary pass 
                 \\firstchoice@false\\else\\firstchoice@true\\fi\\captionold}%
    \\minipage{\\linewidth}% needed else \\vskip in tabular cell causes error
   }{\\endminipage}}
\\makeatother
""",
```

it perturbates less the cells in first tabulary pass, as it inhibits counter stepping (using pre-existing patch done by amsmath to LaTeX counters) rather than suppressing completely the caption (this was my first intention but I had a problem and forgot about it later).

**Comment by jfbu** [comment]
@nmat if you don't need captions you could use `image` directive, as in;

``` rst
Some paragraph text

.. list-table::

   * - .. image:: images/testimage01.png

     - .. image:: images/testimage02.png

   * - .. image:: images/testimage03.png

     - .. image:: images/testimage04.png

```

This will produce LaTeX using `\includegraphics`  which is not problematic inside a LaTeX table/tabular like environment, contrarily to the `figure` environment.

**Comment by tk0miya** [comment]
@jfbu Thank you for comment and great macro :-)
Your macro works almost good. but it does not work inside longtable. (after rewriting to  `\AtBeginEnvironment{longtable}`)

the longtable causes following error:

```
! Misplaced \noalign.
\captionold ->\noalign
                       \bgroup \@ifnextchar [{\egroup \LT@c@ption \@firstofo...
l.178 \caption
              {hello}\label{index:id1}\end{figure}
```

I will investigate the error later :-)

**Comment by jfbu** [comment]
This issue is solved by moving `\let\captionold\caption` to start of longtable environment.

``` latex
\AtBeginEnvironment{longtable}
{\let\captionold\caption
 \renewenvironment{figure}[1][]%
   {\def\@captype{figure}%
....
```

But the problem is that with `tabulary` I exploited the fact that it gives to `\linewidth` a suitable value in the cell, not so for `longtable` (I have not fully checked),  hence there is no obvious choice for the minipage width. Will investigate too :-)

**Comment by jfbu** [comment]
A possible approach would be for longtable to do

``` latex
\AtBeginEnvironment{longtable}
{\let\captionold\caption
 \renewenvironment{figure}[1][]%
   {\def\@captype{figure}%
    \let\caption\captionold
    \begin{minipage}{\PICTUREWIDTH}%
   }{\end{minipage}}}
```

and have the LaTeX writer add suitable `\setlength{\PICTUREWIDTH}{...}` just before the `figure` environment. I guess the LaTeX writer must have possibility to examine width as it may be passed as option ? then it can initialize the `\PICTUREWIDTH` (or a better name !) for that specific picture. We need a `\newlength\PICTUREWIDTH` in sphinx.sty (again with better name!).

Although `longtable` speaks about needing sometimes "multiple passes", this seems to mean "multiple LaTeX compilations", hence perhaps there will not be problem of figure counter incremented twice or more. I tested with a table causing extra LaTeX compilations, but it was ok.

**Comment by jfbu** [comment]
@tk0miya I initially used `\minipage/\endminipage` but it seems ok with `\begin{minipage}/\end{minipage}` as well.

**Comment by jfbu** [comment]
@tk0miya  above I said something misleading about `\linewidth` in `tabulary`. In fact, `tabulary` with `L` columns computes the final cell widths as being proportional to their initial natural width. 

Thus using `\linewidth` in minipage means all pictures appear as having equal widths, independently of their real original widths.

Hence, I think the (hoped-for) method using `\PICTUREWIDTH` should be also applied in case of `tabulary`, else short pictures end up being scaled in case of `\includegraphics{file.png}`. My test had pictures of same widths hence I did not notice initially.

**Comment by tk0miya** [comment]
@jfbu Thank you for great macros. I arrange them and make a PR #2615.
Could you review it please? It allows to specify the width using `:figwidth:` option.

BTW, it would be nice if the physical width of images from the files. Some image formats support to specify DPI values, so it is able to convert px to physical length.
It makes easy to adjust minipage to size of images.

**Comment by jfbu** [comment]
sorry for series of forced updates on my branch showing up above. See https://github.com/sphinx-doc/sphinx/compare/stable...jfbu:figure-in-table which builds upon @tk0miya initial commit in PR #2615.

**Comment by nmat** [comment]
Further information:

I have tested with image instead of figure without the text under it. And it works.

The caption text would be nice to still have as a feature.

**Comment by jfbu** [comment]
@nmat the underlying problem is that `figure` directive uses in LaTeX output the `figure` environment which is a "floating" environment which must be at top level. Using `image` directive means LaTeX will use only the `\includegraphics` command but not wrap it in a `figure` environment. This loses indeed (without extra effort) possibility of a caption. It seems with @tk0miya we are close to a work-around which will allow to Sphinx LaTeX to accept figures put in tables, and still handle correctly captions. But at current time best with these additions to Sphinx LaTeX writer is to add `:figwidth:` option to specify explicitely desired width of figure to be included in table (this width will be obeyed by caption too, possibly becoming multi-line if short).

**Comment by nmat** [comment]
@jfbu that is really good news. Let me know if there is anything I should test.

I am willing to help with testing to see if I can create scenarios for this type of usage with figures.

**Comment by tk0miya** [comment]
Now I merged #2615 now. It will be fixed in next stable release (I will work it today!).
Please reopen if you get same trouble in next stable release.

Thank you for reporting!
